### ISSUE TITLE
Automatic load balancing on Windows has some issues

### ISSUE THREAD

**Issue by randomascii** [ninja-build/ninja#1279]
I've started using -j 1000 -l 42 in order to get optimal build speeds both with and without ninja. However the load balancing calculations in ninja seem to have some problems. With -l 42 (on my 48-thread work station) the number of processes during goma builds sometimes drops very low for what seems like inexplicable reasons. With -l 46 the number of processes during non-goma builds sometimes jumps as high as 600-700, leading to OOM memory errors. Some of this may be fundamental limitations (for perfect results the load balancer needs to predict the future) but I think there are some other factors as well. I have recorded some ETW traces (with an instrumented version of ninja) to better understand what is happening and I am going to document what I find here.

In the screen-show below the blue dots (Id == 61) represent calls to CalculateProcessorLoad where the ticks were updated since the previous call, whereas the pink dots represent calls where the ticks were not updated. So, on 26 of the 40 calls during this 205 ms time period there was no new data. This suggests that new data is only made available infrequently, probably every 15.625 ms or every 20 ms (despite the fact that the timer interrupt frequency is set to 1 ms). Ah, Windows. Also not the blue graph at the bottom which shows idle CPU time which is *extremely* low, averaging just 0.038% - 3.73 CPU ms across the entire trace.

![image](https://cloud.githubusercontent.com/assets/10800041/26080825/158e76b4-397d-11e7-8394-933641fef5ca.png)

In another screen shot we can see some other anomalies. I have selected the first seven events where we got data and you can see them on the timeline. Notice that the idle_ticks_delta is zero, even though the idle thread is running a little bit. However, despite the load_since_last_call (load_recent) being 1.0 the load value increases only very slowly - perhaps the 0.9/0.1 weighted average reacts too slowly given how frequently we don't actually get new data.

![image](https://cloud.githubusercontent.com/assets/10800041/26081089/20f2ccb6-397e-11e7-80e9-d870aa8e57f6.png)

**Comment by nico** [comment]
Yup, see discussion in https://github.com/ninja-build/ninja/pull/798#discussion-diff-15436494 -- "Let's leave it as is for now then, and see how it does in practice." It's almost certainly possible to come up with something that works better. Patches welcome :-)

**Comment by randomascii** [comment]
Thanks for the link. Agreed that no action may be needed. Once I backed off from 46/48 for a load setting to 42/48 it seems to have worked well. The remaining imperfections appear to be from dependencies rather than from load balancing (see crbug.com/722617).
